### ISSUE TITLE
HTTP/2 Push only works when site is not cached

### ISSUE THREAD

**Issue by roebert** [szepeviktor/fix-w3tc#464]
At first my environment:

- Apache 2.4.25
- PHP 7.1.3 mod_php
- MariaDB 10.0.29-22.1
- WordPress 4.7.3
- W3 Total Cache(fixed) 0.9.5.2.4

Page Cache Method: Disk: Enhanced
Minify enabled fpr css/js and HTTP/2 push enabled

when purging all cache and requesting a site for the first time it delivers js/css with http/2-push

![http2push1](https://cloud.githubusercontent.com/assets/6573285/24907037/b1676330-1eba-11e7-9f32-67a468d89383.jpg)

![http2push2](https://cloud.githubusercontent.com/assets/6573285/24907044/bc1bb6dc-1eba-11e7-99bb-53c8c9c0a7af.jpg)

after the site is in the cache it wont http/2-push js/css anymore

![http2push3](https://cloud.githubusercontent.com/assets/6573285/24907327/fc561cb4-1ebb-11e7-9500-42436dbba2e9.jpg)

![http2push4](https://cloud.githubusercontent.com/assets/6573285/24907335/0697c786-1ebc-11e7-9b33-0553b5fa6649.jpg)

maybe i am missing something or did something wrong - i just hope i can help with this issue

**Comment by nigrosimone** [comment]
hi @roebert sorry for the delay.

Check (after a page is in the cache)  if in the `wp-content/cache/page_enhanced/www.yoursite.com/your-cached-page/` folder exists a `.htaccess` file. In the `.htaccess` are stored the list of the assets to push, eg.:

```
<IfModule mod_headers.c>
    Header add Link '</wp-content/cache/minify/1/ebbd6.js>; rel=preload; as=script'
    Header add Link '</wp-content/cache/minify/1/a0a3c.js>; rel=preload; as=script'
    Header add Link '</wp-content/cache/minify/1/773a9.css>; rel=preload; as=style'
</IfModule>
```
this feature require `mod_headers`.

Maybe your Apache doesn't load the module `mod_headers` or doesn't write the `.htaccess` or write it without these headers...

related pr https://github.com/szepeviktor/w3-total-cache-fixed/pull/433

**Comment by roebert** [comment]
hello,

i have checked this - after a purge i requested a site and it was cached 
but no .htaccess was written

```
.../wp-content/cache/page_enhanced/test.example.de/utility-spezial # ls -alF
drwxr-xr-x 2 wwwrun www  4096 Apr 13 10:46 ./
drwxr-xr-x 5 wwwrun www  4096 Apr 13 10:46 ../
-rw-r--r-- 1 wwwrun www 60747 Apr 13 10:46 _index_ssl.html
-rw-r--r-- 1 wwwrun www 12551 Apr 13 10:46 _index_ssl.html_gzip
```

apache has mod_headers

**Comment by nigrosimone** [comment]
@roebert sorry for the long delay.

Your issue require a deep investigation, if you can, i need a detailed log. Before starting **take very careful, you can break your site**.

1. Open your `/wp-config.php` and add or edit these lines, make it `true`:

```php
define('WP_DEBUG', true);
define('WP_DEBUG_LOG', true);
```

Now wordpress puts all error into a log in `/wp-content/debug.log`

2. Open your `/wp-content/plugins/w3-total-cache/Cache_File_Generic.php` go to line 45 and after 

```php
function set( $key, $var, $expire = 0, $group = '' ) {
```

add

```php
    error_log('is_apache:'.json_encode(Util_Environment::is_apache()));
    error_log('headers:'.json_encode($var['headers']));
```

the final code will be:
```php
function set( $key, $var, $expire = 0, $group = '' ) {
    error_log('is_apache:'.json_encode(Util_Environment::is_apache()));
    error_log('headers:'.json_encode($var['headers']));
```

3. Open one page of your website

4. Open `/wp-content/debug.log`, the expected content is something like:

```
[20-Apr-2017 19:08:52 UTC] is_apache:true

[20-Apr-2017 19:08:52 UTC] headers:{"Status-Code":200,"Content-Type":" text\/html; charset=UTF-8","0":{"n":"Link","v":" <http:\/\/www.wordpress.local\/wp-json\/>; rel=\"https:\/\/api.w.org\/\""},"1":{"n":"Link","v":" <http:\/\/www.wordpress.local\/?p=4144>; rel=shortlink"},"2":{"n":"Link","v":" <\/wp-content\/cache\/minify\/1\/ebbd6.js>; rel=preload; as=script"},"3":{"n":"Link","v":" <\/wp-content\/cache\/minify\/1\/c731c.js>; rel=preload; as=script"},"4":{"n":"Link","v":" <\/wp-content\/cache\/minify\/1\/773a9.css>; rel=preload; as=style"}}
```

**Please, copy/paste your log here.**

5. Now you can restore your original  `/wp-config.php` and `/wp-content/plugins/w3-total-cache/Cache_File_Generic.php`.

**Comment by roebert** [comment]
hello,

i tried what you suggested and here are the results:

```
[26-Apr-2017 11:51:51 Europe/Berlin] is_apache:true
[26-Apr-2017 11:51:51 Europe/Berlin] headers:{"Status-Code":200,"Content-Type":" text\/html; charset=UTF-8"}
[26-Apr-2017 11:51:51 Europe/Berlin] is_apache:true
[26-Apr-2017 11:51:51 Europe/Berlin] headers:{"Status-Code":200,"Content-Type":" text\/html; charset=UTF-8"}
```

```
[26-Apr-2017 11:51:48 Europe/Berlin] is_apache:true
[26-Apr-2017 11:51:48 Europe/Berlin] headers:{"Status-Code":200,"Content-Type":" text\/html; charset=UTF-8","X-Pingback":" https:\/\/test.example.de\/xmlrpc.php"}
[26-Apr-2017 11:51:48 Europe/Berlin] is_apache:true
[26-Apr-2017 11:51:48 Europe/Berlin] headers:{"Status-Code":200,"Content-Type":" text\/html; charset=UTF-8","X-Pingback":" https:\/\/test.example.de\/xmlrpc.php"}
```

it seems this log-entries only get written when the site is not in cache - i tried with different sites, the information is the same every time: code 200 - content-type:text/html - charset:utf-8

**Comment by nigrosimone** [comment]
@roebert sorry for the delay

what i see into your log is the absense of the "Link rel=preload" headers... see the difference between my and your log :

my:
```json
headers:{  
   "Status-Code":200,
   "Content-Type":" text\/html; charset=UTF-8",
   "0":{  
      "n":"Link",
      "v":" <http:\/\/www.wordpress.local\/wp-json\/>; rel=\"https:\/\/api.w.org\/\""
   },
   "1":{  
      "n":"Link",
      "v":" <http:\/\/www.wordpress.local\/?p=4144>; rel=shortlink"
   },
   "2":{  
      "n":"Link",
      "v":" <\/wp-content\/cache\/minify\/1\/ebbd6.js>; rel=preload; as=script"
   },
   "3":{  
      "n":"Link",
      "v":" <\/wp-content\/cache\/minify\/1\/c731c.js>; rel=preload; as=script"
   },
   "4":{  
      "n":"Link",
      "v":" <\/wp-content\/cache\/minify\/1\/773a9.css>; rel=preload; as=style"
   }
}
```
your:
```json
headers:{  
   "Status-Code":200,
   "Content-Type":" text\/html; charset=UTF-8"
}
```

without `Link`, http2 can't push the assets.

have you checked during this test, into the minify page setting for css and js, the http2push option?

**Comment by roebert** [comment]
Hello,

minify for css and js is enabled
http2push is enabled for css/js

i've checked it again - first request for a page after complete purge: js and css came with preload
second request: no preload headers

like you wrote before, no .htaccess is written in the cache folders

**Comment by nigrosimone** [comment]
@roebert sorry for the very long delay.

I have fixed an issue and maybe it can help you: https://github.com/szepeviktor/w3-total-cache-fixed/pull/491

**Comment by roebert** [comment]
i deployed a whole new clone of the site and tried it again - the problem is gone.
.htaccess is written for every page and the preload header works

only the start-page wont create a .htaccess with preload options

**Comment by nigrosimone** [comment]
> only the start-page wont create a .htaccess with preload options

yes this is a side effect of https://github.com/szepeviktor/w3-total-cache-fixed/pull/491 it fix some bugs but disable http 2 push for only the root/home page
